lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,282 @@
+//===- DecorateWithLayoutInfoSPIRVPass.cpp - Decorate with layout info ----===//
+//
+// Copyright 2019 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file implements a pass to decorate SPIR-V types with layout info.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SPIRV/Passes.h"
+#include "mlir/Dialect/SPIRV/SPIRVDialect.h"
+#include "mlir/Dialect/SPIRV/SPIRVOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Support/StringExtras.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+class SPIRVLayoutInfoUtils {
+public:
+  /// Return a new spirv::StructType with layout info, increment an offset by
+  /// proccessed type size in bytes.
+  static Type decorateType(spirv::StructType structType,
+                           spirv::StructType::LayoutInfo &offset);
+  /// Check that type is legal in terms of SPIR-V layout info decoration.
+  static bool isLegalType(Type type);
+
+private:
+  static Type decorateType(Type type, spirv::StructType::LayoutInfo &offset);
+  static Type decorateType(spirv::CompositeType compositeType,
+                           spirv::StructType::LayoutInfo &offset);
+  static Type decorateType(spirv::PointerType pointerType,
+                           spirv::StructType::LayoutInfo &offset);
+  /// Create an array or a vector type with layout info, based on composite
+  /// type kind.
+  static Type getArrayOrVectorType(spirv::CompositeType compositeType,
+                                   Type elementType, uint64_t arrayStride = 0);
+  /// Calculate the size in bytes of the primitive type.
+  static unsigned getPrimitiveTypeSizeInBytes(Type primitiveType);
+  /// Check that type is primitive in terms of SPIR-V dialect.
+  static bool isPrimitiveType(Type type);
+};
+
+Type SPIRVLayoutInfoUtils::decorateType(spirv::StructType structType,
```

Comment:
This method needs to account for alignment of the data types as well. For example if you have an spv.struct<f32, f64>, then you need to
1) Align the entire struct to the largest alignment of any of its members
2) Add padding so that individual members are aligned.

I have to look at the SPEC carefully, but definitely needs taking care of alignment here.

Reply:
@MaheshRavishankar oh, I've missed the alignment, thanks for mention this, you are right! I'll fix it.